Removed Error Log to Avoid Spam Logging

Marcosrico opened the pull request:
Deleted logging error in `ZkClient` delete method to eliminate log noise.

### Issues

https://github.com/apache/helix/issues/2202#issue-1357887240
References an issue where despite a successful retry, failure to remove a node still throws an error log.

### Description

https://github.com/apache/helix/issues/2202#issue-1357887240

Removed a failed remove log in `ZkClient.delete()`. This log error shows when there is an attempt to delete a parent ZNode with non-empty children. However, it automatically calls another function that recursively deletes the children of the node and in the case this operation is successful, having a stack-trace error from before can confuse users of a problem that is in fact trivial. By deleting the log shown when the above error occurs, we eliminate these misleading logs and avoid log flooding. We can do this as the exception is already being thrown so it is not necessary to log it. 

### Tests

Added `testDeleteNodeWithChildren()` in `TestZkBaseDataAccessor.java`

The test creates a 10 instance message records. It proceeds to call remove on the parent of those children. The test checks whether remove returns true, which it does. However, the purpose of it is to simulate the initial error that brought upon this change. Going through the debugger, we see that it does go through the right path and effectively does not throw any logs (as it has been removed).

```
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 57.175 s - in org.apache.helix.manager.zk.TestZkBaseDataAccessor
[INFO] Results:
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```



Local test result for failed CI tests:
For `testZKHelixManager` and `afterClass`

```
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 208.496 s - in org.apache.helix.integration.multizk.TestMultiZkConectionConfig
[INFO] Results:
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
```

For `testZkClientCreationMultiZk`

```
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 79.998 s - in org.apache.helix.task.TestTaskStateModelFactory
[INFO] Results:
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```

qqu0127 commented:
Hi Marcos, one quick thing, we don't post internal ticket to github, it's not accessible to outside world anyway. Please use github issue instead.

xyuanlu commented on helix-core/src/test/java/org/apache/helix/manager/zk/TestZkBaseDataAccessor.java:
```diff
@@ -433,6 +433,29 @@ public void testSyncRemove() {
+    Assert.assertTrue(accessor.remove(PropertyPathBuilder.instanceMessage(root, "host_1"), 0),
+            "Should return True despite log errors.");
+  }
+
```
Thanks for making this change!! 
I would suggest validate that these ZNode has already being deleted.

xyuanlu commented on helix-core/src/test/java/org/apache/helix/manager/zk/TestZkBaseDataAccessor.java:
```diff
@@ -433,6 +433,37 @@ public void testSyncRemove() {
+  @Test
+  public void testDeleteLogging() {
+    String root = _rootPath;
+    _gZkClient.deleteRecursively("/" + root);
```
I might have missed it last time. Why we are deleting everything under root?

xyuanlu replied:
Also please link the issue number if the description.

Marcosrico replied:
This is to make sure we work with a clean path. However, it might not be necessary now that you point it out.

junkaixue commented:
NIT: please remove original template context in the details.

qqu0127 commented on helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java:
```diff
@@ -726,7 +726,7 @@ public boolean remove(String path, int options) {
       _zkClient.delete(path);
     } catch (ZkException e) {
-      LOG.debug("Failed to delete {} with opts {}, err: {}. Try recursive delete", path, options,
+      LOG.info("Failed to delete {} with opts {}, err: {}. Try recursive delete", path, options,
```
Just FYI, change from debug to info is a up-level. We may see more logs like this, is this what we want? (most default log level is at INFO or plus)

Marcosrico replied:
Hmm you're right. What we're looking for is to diminish logging clutter - increasing the logging level would worsen that! I wanted to have the logging level consistent with the log error in the remove function but that wouldn't make sense to the overall objective of this fix. Thanks for the comment!

xyuanlu replied:
The aim for this change is to make the log less misleading. By increasing the log level will indicate user that the previous warning log saying failed to remove znode won't be an issue and we will retry remove recursively. 
This is only in failure path and won't be causing log flooding.

rahulrane50 replied:
neat : I understand the overall intention of changing this to INFO level but I would be very cautious about this change since this is very core method which is used in many places. Any small changes in logging here would flood logs in specific scenarios. Also could you help me understand why was the "debug" log noise in the first place. It shouldn't even show up in logs unless debug logging is enabled right? (I assume debug logging is disabled by default)

Marcosrico replied:
I am not too familiar with what the default logging is for the different applications of these methods. However, the ticket and issue created arose from customers (i.e. Venice) who have run into the issue of these "noisy" logs. Ideally, we would want to completely avoid calling these logs altogether (in the case that deletion succeeds further on). However, we weren't able to come up with a liable solution. Thus, if the log error is to be shown regardless, the purpose of this change as Xiaoyuan mentioned is to make it less misleading. Let me know what you think!

jiajunwang replied:
I'm confused here.
If set the log level to info, then orignal code will NOT print debug log, but the new change will be printing more logs. It contradicts to the description.

> A log request of level p in a logger with level q is enabled if p >= q. This rule is at the heart of log4j. It assumes that levels are ordered. For the standard levels, we have ALL < DEBUG < INFO < WARN < ERROR < FATAL < OFF.
Here, in the original code p == debug and q == info. So the log won't be printed.

Debug to Info is considered upgrade instead of downgrade.

Marcosrico replied:
You're right, I made a mistake in the description, thank you for pointing that out! I'll update the description with this feedback.

qqu0127 commented on helix-core/src/test/java/org/apache/helix/manager/zk/TestZkBaseDataAccessor.java:
```diff
@@ -433,6 +433,36 @@ public void testSyncRemove() {
   }
 
+  @Test
+  public void testDeleteLogging() {
+    String root = _rootPath;
+
+    ZkBaseDataAccessor<ZNRecord> accessor = new ZkBaseDataAccessor<>(_gZkClient);
```
Thanks for writing up the test. 
Logging test can be tricky, it depends on the configuration of log level in test environment, and there isn't a good way to verify the log message. 
If I understand correctly, this test is only to test logging, but not to test the delete functionality?

Marcosrico replied:
Correct, this is just to check that the specific logging message is displayed. I believe delete functionality is tested in `TestZkBaseDataAccessor.testAsyncZkBaseDataAccessor()`

qqu0127 replied:
Thanks. In this case, I don't think we need this test at all.

Marcosrico replied:
I might have phrased my answer wrong. The test does check whether zk catches the error of removing a parent node with children, which I believe is not tested elsewhere, though not sure if that is considered as delete functionality. The primary motive for the test is to catch this error and make sure the right logging is made. Perhaps that is not necessary given the variable nature of log configs - let me know, thanks!

qqu0127 replied:
Ok, I see your point. My concern is that, `"make sure the right logging is made"` is checked manually. It doesn't add much value to the unit test in regard to logging. Functionally, it does have value to verify no exception is thrown. So we may keep the test, and rename it to e.g. `testDeleteNodeWithChildren`, just my two cents.

Marcosrico replied:
I agree with renaming the tests for better clarity - thanks for the feedback!

xyuanlu commented:
For line 2069 in [ZkClient.java](https://github.com/apache/helix/pull/2201/files#diff-c20891e1adafdc5bd7c55936d81ec52e84081b76f7df3759da762f932d3f1cae), we should remove 
`if (LOG.isDebugEnabled()) {`
It is redundant.

rahulrane50 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -2073,7 +2073,7 @@ public Object call() throws Exception {
     } catch (Exception e) {
       recordFailure(path, ZkClientMonitor.AccessType.WRITE);
-      LOG.warn("zkclient {}, Failed to delete path {}! ", _uid, path, e);
+      LOG.debug("zkclient {}, Failed to delete path {}! ", _uid, path, e);
```
Umm is this correct? There are 3 parameters for 2 positional brackets.

Marcosrico replied:
I think the first two parameters are for the positional brackets. The last one since it's an exception it is thrown as a separate message. Let me know if I'm mistaken

xyuanlu commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -2073,7 +2073,7 @@ public Object call() throws Exception {
     } catch (Exception e) {
       recordFailure(path, ZkClientMonitor.AccessType.WRITE);
-      LOG.warn("zkclient {}, Failed to delete path {}! ", _uid, path, e);
+      LOG.info("zkclient {}, Failed to delete path {}! ", _uid, path, e);
```
+1 on this.

Marcosrico replied:
That is a very good point, thank you!

jiajunwang commented:
> For line 2069 in [ZkClient.java](https://github.com/apache/helix/pull/2201/files#diff-c20891e1adafdc5bd7c55936d81ec52e84081b76f7df3759da762f932d3f1cae), we should remove `if (LOG.isDebugEnabled()) {` It is redundant.

This is in general to avoid error message computing (if applicable), like format. Not necessarily redundent. But IMO, I think it looks ugly and does not really save much, I vote for removing them : )

xyuanlu commented:
LGTM. Thanks for making the change.

NealSun96 commented:
CI workflows approved.

xyuanlu commented:
> 

Please correct me if I am wrong. The this kind of log formatter, log4j is already doing lazily log construction. So we will only do string construction when the log level<= debug. 
https://logging.apache.org/log4j/2.0/manual/api.html#Java_8_lambda_support_for_lazy_logging

xyuanlu commented:
> > For line 2069 in [ZkClient.java](https://github.com/apache/helix/pull/2201/files#diff-c20891e1adafdc5bd7c55936d81ec52e84081b76f7df3759da762f932d3f1cae), we should remove `if (LOG.isDebugEnabled()) {` It is redundant.
> 
> This is in general to avoid error message computing (if applicable), like format. Not necessarily redundent. But IMO, I think it looks ugly and does not really save much, I vote for removing them : )

Please correct me if I am wrong. I think with the correct log formatter, log4j will do lazily log message construction. We should also do this kind of log improvement for Helix as well....
link:
https://logging.apache.org/log4j/2.x/manual/api.html

Marcosrico commented:
This PR is ready to merge, approved by @xyuanlu 
Failed tests are not stable and pass locally. See description for test results.

Commit message:

Removed Error Log to Avoid Spam Logging
Removed an unnecessary and misleading error log in ZkClient. This cleans up the logging and avoids clutter.
